### PR TITLE
Roster get improved

### PR DESCRIPTION
**Pull request by bartekgorny**
This PR is required by #1286 . The issue being fixed here is that backends to mod roster do not provide an api for finding out whether a given jid is in roster or not - they always return a roster entry, while for rest we sometimes have to check it.

The change is that basic getters for roster return either a roster entry or an atom `does_not_exist`, and it is the caller that decides what to do with it.

### CODE REVIEW

File: apps/ejabberd/src/mod_roster.erl [hunk 1]
```diff
@@ -191,6 +205,26 @@ stop(Host) ->
...
+    mod_roster_backend:get_roster_entry(jid:nameprep(LUser), LServer, jid_arg_to_lower(Jid)).
+
+get_roster_entry(LUser, LServer, Jid, full) ->
+    mod_roster_backend:get_roster_entry(jid:nameprep(LUser), LServer, jid_arg_to_lower(Jid), full).
+
+get_roster_entry_t(LUser, LServer, Jid) ->
+    mod_roster_backend:get_roster_entry_t(jid:nameprep(LUser), LServer, jid_arg_to_lower(Jid)).
+
+get_roster_entry_t(LUser, LServer, Jid, full) ->
+    mod_roster_backend:get_roster_entry_t(jid:nameprep(LUser), LServer, jid_arg_to_lower(Jid), full).
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_roster.erl]: According to **Elvis**:
> Line 218 is too long:     mod_roster_backend:get_roster_entry_t(jid:nameprep(LUser), LServer, jid_arg_to_lower(Jid), full)..

File: apps/ejabberd/src/mod_roster.erl [hunk 2]
```diff
@@ -67,6 +72,10 @@
 
 -type roster() :: #roster{}.
 
+-type sub_presence() :: subscribe | subscribed | unsubscribe | unsubscribed.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_roster.erl]: According to **Compiler**:
> Warning: type sub_presence() is unused

**Comment by bartekgorny** [comment]
Yes, there are some intermediate commits, I'll rebase and squash them.

File: apps/ejabberd/src/mod_muc_light.erl [hunk 3]
```diff
@@ -285,11 +285,14 @@ remove_user(Acc, User, Server) ->
...
--spec add_rooms_to_roster(Acc :: [mod_roster:roster()], UserUS :: ejabberd:simple_bare_jid()) ->
-    [mod_roster:roster()].
+-spec add_rooms_to_roster(Acc :: mongoose_acc:t(), UserUS :: ejabberd:simple_bare_jid()) ->
+    mongoose_acc:t().
 add_rooms_to_roster(Acc, UserUS) ->
-    lists:foldl(
-      fun({{RoomU, RoomS}, RoomName, RoomVersion}, Acc0) ->
+    Items = mongoose_acc:get(roster, Acc, []),
+    RoomList = mod_muc_light_db_backend:get_user_rooms(UserUS, undefined),
+    Info = get_rooms_info(lists:sort(RoomList )),
```
Comment (rslota) [inline apps/ejabberd/src/mod_muc_light.erl]: Just a random whitespace :)

File: apps/ejabberd/src/mod_roster.erl.orig [hunk 4]
```diff
@@ -0,0 +1,929 @@
+%%%----------------------------------------------------------------------
```
Comment (rslota) [inline apps/ejabberd/src/mod_roster.erl.orig]: I guess this file is to be removed.

File: apps/ejabberd/src/mod_roster_odbc.erl [hunk 5]
```diff
@@ -225,16 +245,18 @@ read_subscription_and_groups(LUser, LServer, LJID) ->
                                <<"F">> -> from;
                                _ -> none
                            end,
-            Groups = case catch
-                          rdbms_queries:get_rostergroup_by_jid(LServer, Username,
-                                                              SJID)
+            Groups = case catch rdbms_q(GRFunc, [LServer, Username, SJID])
+%%                          rdbms_queries:get_rostergroup_by_jid(LServer, Username,
```
Comment (rslota) [inline apps/ejabberd/src/mod_roster_odbc.erl]: Is this code used? If not why don't we remove it?

File: test.disabled/ejabberd_tests/rebar.config [hunk 6]
```diff
@@ -13,7 +13,7 @@
         {exml, ".*", {git, "git://github.com/esl/exml.git", "d365533"}},
         {escalus, ".*", {git, "git://github.com/esl/escalus.git", "868afbb"}},
         %% Switch cowboy to upstream after ditching support for OTP 17.x
-        {cowboy, ".*", {git, "git://github.com/rslota/cowboy.git", {tag, "2.0.0-pre.7-r17"}}},
+        {cowboy, ".*", {git, "git://github.com/bartekgorny/cowboy.git", {branch, "quickfix"}}},
```
Comment (rslota) [inline test.disabled/ejabberd_tests/rebar.config]: It's not required any more, you can revert this change.
